fuzz/fuzz_targets/roundtrip-wit.rs

```diff
@@ -0,0 +1,719 @@
+    // them and insert them into a `new_resolve`.
+    let mut new_deps = HashMap::new();
+    let mut new_resolve = Resolve::default();
+    let mut last = None;
+    for (_, pkg) in resolve.packages.iter() {
+        let mut map = SourceMap::new();
+        let pkg_name = &pkg.name;
+        for (name, doc) in pkg.documents.iter() {
+            let doc = DocumentPrinter::default().print(resolve, *doc).unwrap();
+            write_file(&format!("{file}-{pkg_name}-{name}.wit"), &doc);
+            map.push(format!("{name}.wit").as_ref(), &name, doc);
+        }
+        let unresolved = map.parse(&pkg.name, pkg.url.as_deref()).unwrap();
+        let id = new_resolve.push(unresolved, &new_deps).unwrap();
+        new_deps.insert(pkg.name.clone(), id);
+        last = Some(id);
+    }
+
+    // Finally encode the `new_resolve` which should be the exact same as
+    // before.
+    let wasm2 = wit_component::encode(&new_resolve, last.unwrap()).unwrap();
+    write_file(&format!("{file}-reencoded.wasm"), &wasm2);
+    if wasm != wasm2 {
+        panic!("failed to roundtrip through text printing");
+    }
+
+    wasm
+}
+
+fn write_file(path: &str, contents: impl AsRef<[u8]>) {
+    if !log::log_enabled!(log::Level::Debug) {
+        return;
+    }
+    log::debug!("writing file {path}");
+    let contents = contents.as_ref();
+    let path = Path::new(path);
+    std::fs::write(path, contents).unwrap();
+    if path.extension().and_then(|s| s.to_str()) == Some("wasm") {
+        let path = path.with_extension("wat");
+        log::debug!("writing file {}", path.display());
+        std::fs::write(path, wasmprinter::print_bytes(&contents).unwrap()).unwrap();
+    }
+}
+
+mod generate {
+    use arbitrary::{Arbitrary, Result, Unstructured};
+    use std::collections::HashSet;
+    use std::mem;
+    use std::str;
+    use wit_parser::*;
+
+    const MAX_PARTS: usize = 5;
+
+    #[derive(Default)]
+    struct Generator {
+        packages: PackageList,
+        documents: DocumentList,
+        interfaces: InterfaceList,
+        next_interface_id: u32,
+    }
+
+    type TypeList = Vec<(String, usize)>;
+    type InterfaceList = Vec<(String, u32, bool, TypeList)>;
+    type DocumentList = Vec<(String, InterfaceList)>;
+    type PackageList = Vec<(String, DocumentList)>;
+
+    struct InterfaceGenerator<'a> {
+        gen: &'a Generator,
+        unique_names: HashSet<String>,
+        types_in_interface: Vec<(String, usize)>,
+    }
+
+    pub struct Package {
+        pub name: String,
+        pub sources: SourceMap,
+    }
+
+    pub fn packages(u: &mut Unstructured<'_>) -> Result<Vec<Package>> {
+        Generator::default().gen(u)
+    }
+
+    impl Generator {
+        fn gen(&mut self, u: &mut Unstructured<'_>) -> Result<Vec<Package>> {
+            let mut packages = Vec::new();
+            let mut names = HashSet::new();
+            while packages.len() < 10 && (packages.is_empty() || u.arbitrary()?) {
```

Comment:
Nit: can we add constants for these maximums (here and in a few other places in c42cfd2ab471e20dc78812131e05c3211d5e474a) to go along side `MAX_PARTS`? It just makes it a little easier to understand that these are limits being placed intentionally on what is generated.